app/lib/otis/log_importer.rb

```
@@ -0,0 +1,163 @@
+            source_path: log_struct["Path"],
+            destination_directory: tempdir
+          )
+          Rails.logger.info "processing log #{temp_file} from #{log_struct["Path"]}"
+          process_file temp_file
+          File.unlink temp_file
+          @stats[:files_scanned] += 1
+        end
+      end
+      write_journal
+      self
+    end
+
+    # Returns Time object, either the last time the process was run,
+    # or the Epoch to represent "distant past"
+    def last_import
+      @last_import ||= if File.exist?(journal_path)
+        File.open(journal_path) do |file|
+          Time.parse file.gets.chomp
+        end
+      else
+        Time.at 0
+      end
+    end
+
+    def write_journal
+      File.open(journal_path, "w") do |file|
+        file.puts transfer.query_time
+      end
+    end
+
+    # For now this file will be the same `babel/logs` location as its predecessor
+    def journal_path
+      @journal_path ||= File.join(journal_directory, LOG_IMPORT_JOURNAL_NAME)
+    end
+
+    def journal_directory
+      @journal_directory ||= begin
+        dir = ENV.fetch("OTIS_LOG_IMPORT_JOURNAL_DIRECTORY", DEFAULT_JOURNAL_DIRECTORY)
+        if !File.writable?(dir)
+          raise "journal directory #{dir} is not writable"
+        end
+        dir
+      end
+    end
+
+    def process_file(log_file)
+      Zinzout.zin(log_file, encoding: "utf-8") do |infile|
+        infile.each_with_index do |line, line_number|
+          begin
+            entry = JSON.parse(line)
+          rescue JSON::ParserError
+            Rails.logger.error "unparseable JSON in #{log_file} line #{line_number + 1}"
+            next
+          end
+          if relevant_log_entry?(entry)
+            # Call `.save` and not `.save!` on this object because its SHA may
+            # violate uniqueness, which is perfectly okay.
+            report = create_report(entry)
+            @stats[:entries_found] += 1
+            if report.save
```

Would it clean this up if we swallowed the exception if there's a duplicate key error based on the sha, and log it otherwise? Failing that, maybe we should extract this logic to a separate method?

---

Unfortunately if any validation fails (from `save!`) it seems to raise `ActiveRecord::RecordInvalid` and I have to look at the errors anyway. I am moving this down out of the main loop however.